tools/ontology-builder/src/compute_descendent_mappings.py

```diff
@@ -0,0 +1,279 @@
+    :param tissue_type: str tissue type
+    :return: str entity name with tissue type appended
+    """
+    # Tissue types
+    tissue_type_cell_culture = "cell culture"
+    tissue_type_organoid = "organoid"
+
+    # Handle error case (possible if tissue has not been migrated to 4.0.0+ schema).
+    if tissue_type is None:
+        return entity_name
+
+    if tissue_type == tissue_type_cell_culture:
+        # true if the given tissue type is "cell culture".
+        return f"{entity_name} ({tissue_type_cell_culture})"
+
+    if tissue_type == tissue_type_organoid:
+        # true if the given tissue type is "organoid".
+        return f"{entity_name} ({tissue_type_organoid})"
+
+    return entity_name
+
+
+def key_organoids_by_ontology_term_id(entity_names: List[str]) -> Dict[str, str]:
+    """
+    Returns a dictionary of organoid ontology term IDs by stem ontology term ID.
+
+    :param entity_names: List of entity names
+    :return: Dict of organoid ontology term IDs by ontology term ID
+    """
+
+    organoids_by_ontology_term_id = {}
+    for entity_name in entity_names:
+        if "(organoid)" in entity_name:
+            """
+            Historically (i.e. before schema 4.0.0 and the introduction of
+            `tissue_type`), tissues of type "organoid" were tagged with "(organoid)"
+            in their labels and ontology IDs. The post-4.0.0 `tissue_type` value is
+            mapped to this tagged version in order to minimize downstream updates to
+            the filter functionality.
+            """
+            ontology_term_id = entity_name.replace(" (organoid)", "")
+            organoids_by_ontology_term_id[ontology_term_id] = entity_name
+
+    return organoids_by_ontology_term_id
+
+
+def build_descendants_by_entity(
+    entity_hierarchy: List[List[str]], ontology_parser: OntologyParser
+) -> Dict[str, List[str]]:
+    """
+    Create descendant relationships between the given entity hierarchy.
+
+    :param entity_hierarchy: List of lists of entity names
+    :param ontology_parser: OntologyParser instance
+    :return: Dict of descendants by term_id
+    """
+    all_descendants = {}
+    for idx, entity_set in enumerate(entity_hierarchy):
+        # Create the set of descendants that can be included for this entity set.
+        # For example, systems can include organs or tissues,
+        # organs can only include tissues, tissues can't have descendants.
+        accept_lists = entity_hierarchy[idx + 1 :]
+
+        # Tissue or cell type for example will not have any descendants.
+        if not accept_lists:
+            continue
+
+        accept_list = [i for sublist in accept_lists for i in sublist]
+        organoids_by_ontology_term_id = key_organoids_by_ontology_term_id(accept_list)
+
+        # List descendants of entity in this set.
+        for entity_name in entity_set:
+            descendants = set(ontology_parser.get_terms_descendants(entity_name)[entity_name])
```

Comment:
why not pass in entity_set to get_terms_descendants and then work with the resulting map?

Reply:
I moved this call outside of this loop.